stan-dev/math
# Untangles all of our includes!

#### Summary:

This pull request untangles all of our includes. 

For prim vs rev vs fwd:
- prim is standalone and does not include rev or fwd
- rev includes prim, but not fwd
- fwd includes prim, but not rev
- mix includes prim, rev, and fwd

For scal vs arr vs mat:
- scal is standalone
- arr includes scal, but not mat
- mat includes arr and scal

Most of this was just shifting code around. There were two major pieces that I had to rewrite in order to get this to work:
- VectorView
- OperandsAndPartials.

All tests still pass. This might reduce compile times a bit.

#### Intended Effect:

Detangle all our headers! Makes the math library easier to use. We can now maintain this going forward.

#### How to Verify:

Run tests. See VectorView and OperandsAndPartials to see if that makes sense. I added some more doc for those.

#### Side Effects:

Compile times might be slightly shorter!

#### Documentation:

Included.

#### Reviewer Suggestions: 

Anyone.

P.S. Sorry for the really large number of files.
